### ISSUE TITLE
Add a search function to the projects page

### ISSUE THREAD

**Issue by jdingeman** [hackforla/website#4136]
### Overview
The [projects page](https://www.hackforla.org/projects/) needs a search box so that a user can search all of the projects by keywords. This issue is to implement the final design for the search box. 

### Action Items
- [x] Add a search box at the top of the projects to reflect the design in Figma
- [x] Similar to a user using ctrl + f to find a word on a webpage, the user can search any word or words from the descriptions of the projects
  - [x] The difference would be that the search function will narrow down the projects rendered on the page based on user input
- [x] Include functionality for [boolean operators](https://southern.libguides.com/google/boolean) for users to refine their search
  - [x] Only use operators AND, OR and - 
   <html>
<body>

Boolean Operators | Meaning | Example
-- | -- | --
**AND** | Limit results | Vegetarianism AND obesity(Search for content that contains both vegetarianism and obesity.)
**OR** | One term OR another | Scholarly OR academic(Search for content that contains scholarly OR academic.)
**-** | Exclude a term from the search | Bears -Chicago(Limits results to only those with bears and not the term Chicago.)

</body>
</html>

- [x] Make the appropriate changes to the CSS, HTML, and JS as you need to
- [x] The goal is to make this component reusable on multiple pages so it can be used for current and future pages
- [x] Once you have submitted the PR, please write up what files go into the search bar component and any other details that a person adding the same component to another page (e.g., toolkit page) will need to know. You can leave that write up here in this issue as a comment.

### Resources/Instructions
- See the [final figma design](https://www.figma.com/file/0RRPy1Ph7HafI3qOITg0Mr/Hack-for-LA-Website?node-id=24096-281035&t=jNu8aChcchEtRwmG-0)
   - see [toolkit mobile figma latest](https://www.figma.com/file/0RRPy1Ph7HafI3qOITg0Mr/Hack-for-LA-Website?type=design&node-id=24535-287271&mode=design&t=f7zbZLTg6r3V9B07-0) for search inspo 
- In the [_includes](https://github.com/hackforla/website/blob/gh-pages/_includes/) page you should find the [current-projects.html](https://github.com/hackforla/website/blob/gh-pages/_includes/current-projects.html)
- In the [assets/js](https://github.com/hackforla/website/blob/gh-pages/assets/js) folder you should find [project.js](https://github.com/hackforla/website/blob/gh-pages/assets/js/project.js)
- [Standardized Components](https://github.com/hackforla/website/wiki/Standardized-Components) wiki
-

**Comment by ExperimentsInHonesty** [comment]
@jdingeman I don't think the project cards have any keywords.  So It should search all the words on the cards (which is something the filters don't do.  I realize people could just use the browser find, but this will also narrow down the results to things that match.   Can you rephrase?  Also, do we want to give it some acceptable boolean operators?  AND, OR, etc.? https://southern.libguides.com/google/boolean

<html>
<body>


Boolean Operators | Meaning | Example
-- | -- | --
**AND** | Limit results | Vegetarianism AND obesity(Search for content that contains both vegetarianism and obesity.)
**OR** | One term OR another | Scholarly OR academic(Search for content that contains scholarly OR academic.)
**-** | Exclude a term from the search | Bears -Chicago(Limits results to only those with bears and not the term Chicago.)


</body>
</html>

**Comment by jdingeman** [comment]
@ExperimentsInHonesty - that's a great idea. I added that in. I wonder if this would end up making it a complexity: large issue. What do you think?

**Comment by ExperimentsInHonesty** [comment]
@kurikurichan You have not posted your ETA on this issue.  If we don't hear from you by Sunday's meeting, we will assume you are not working on it and return it to the prioritized backlog.

**Comment by Skydodle** [comment]
Availability: Tues Thurs Sat 6-9pm PST
ETA: I have learned to give myself at least two weeks on large issue, so tentative ETA 4/10 EOD

**Comment by Skydodle** [comment]
Progress: I had to postpone working on this issue due to a couple of take home interviews. Will start working on this issue today and try to wrap up ASAP
Blockers: None
ETA: 4/17 EOD
Availability: M-F 6 to 9pm PST

**Comment by ExperimentsInHonesty** [comment]
@Skydodle if you have worked on the issue recently, please provide an update.  Otherwise please unassign yourself and put the issue in the questions and review column with a comment.

**Comment by Skydodle** [comment]
Progress: Apologies for the significant delay. I have been occupied with unexpected events. I'm going to spend as much time as possible this week to complete this issue.
Blockers: No blockers
Availability: Everyday from today until May 23
ETA: Next Tuesday May 23

**Comment by Skydodle** [comment]
Apologies, I am unable to work on this issue due to recent personal events. Thank you for your patience and understanding.

**Comment by ExperimentsInHonesty** [comment]
@hackforla/website-merge Please review, cleanup and add back to prioritized backlog

**Comment by Adastros** [comment]
Hi @ExperimentsInHonesty , pull request #4694 updated the projects filter so that the tags are now on top of the project list where the search bar should be. There was some confusion on the filter tag location.

To confirm, the original intention of the [proposed solution](https://github.com/hackforla/website/issues/4530#issuecomment-1521032707 ) was to fix the scrolling issue without changing the behavior of the filter tags correct? 

- If yes, should a new issue be created to move the filter tags back to their original location before this issue can be worked on? 
- If no, is there a preference where the search bar should be located since the updates made by this won't match the Figma design?  

<details>
<summary>Filter Tag Original location</summary>

![image](https://github.com/hackforla/website/assets/48114590/8275dafa-2215-4ed2-8aef-efcb2678b30b)

</details>

<details>
<summary>Updated Filter Tag Location</summary>

![image](https://github.com/hackforla/website/assets/48114590/8d957be8-a148-4edc-ba81-41b863981045)

</details>

**Comment by t-will-gillis** [comment]
@Adastros's comment https://github.com/hackforla/website/issues/4136#issuecomment-1558599739 is correct. The red filter tags should move back to their original (and intended) placement to prevent conflict with the search bar- this is being addressed by https://github.com/hackforla/website/issues/4753

**Comment by Adastros** [comment]
Reviewed and moving back to prioritized backlog.

**Comment by Thinking-Panda** [comment]
Availability: M-F 10 AM -3PM
ETA: July 7

**Comment by Thinking-Panda** [comment]
Progress: I have worked on the HTML and CSS part. I am working on the Js file.
Blockers: Had trouble debugging on VS Code and now I am debugging on the browser. It kind of slows me down as docker takes time to reload
Availability: M-F 2-3 hrs
ETA: 7/12
Pictures (optional):
![Mobile View](https://github.com/hackforla/website/assets/122488603/4dac5556-4b9c-435d-96a2-d216a6471d18)
![TabletView](https://github.com/hackforla/website/assets/122488603/3ce29a3c-b70c-4681-a881-83706d80b967)
![DesktopView](https://github.com/hackforla/website/assets/122488603/bec23e02-01bb-46b0-8f92-b665f0be3f49)

**Comment by ronaldpaek** [comment]
The more I think about the search box I am wondering is this feature going to be used more for desktop users or mobile? I don't think mobile users will want to type that much, and if they do is it going to just be fuzzy searching? Cause it seems like the updates the UI everytime the URL value changes and it's probably parsing the data very specifically.

Is keyword search meant to be something that's suppose to be robust and have many combinations plus also work with the current checkboxes as in append to the end of it? and vice versa are they supposed to work in sync and add a checkbox and remove a checkbox if I reference it again??

I think u could have the output of the current search params display on the search by keywords bar and you can add stuff to id I think that would be feasible, but I don't know how visually appealing that would be?

<img width="598" alt="CleanShot 2023-07-13 at 02 21 16@2x" src="https://github.com/hackforla/website/assets/95109313/3e919049-b354-440d-acab-4f475aad4b26">

@t-will-gillis @roslynwythe

**Comment by Thinking-Panda** [comment]
**Documentation:**
- Add the form with the search textbox and buttons to the webpage's `html` file.
- Style properties for the form is coded in `_sass\elements\_search-bar.scss` which is already imported into `main.scss`
- Changes made to the `Js` file include adding events related to the search bar and adding necessary functions to get the expected results.
	- 4 events added are focus and keydown events for the textbox, click events for the buttons.
	- new functions added are `searchOnFocusEventHandler()`,` searchEnterKeyHandler(e)`, `searchEventHandler()`, `searchCloseEventHandler()`.
	- Changes were also made to `checkBoxEventHandler()` to update search/filter parameters, `updateCategoryCounter()`, `updateProjectCardDisplayState()` to add search functionality with different operators, `projectCardComponent()` to include more `data-*` attribute to enable search in specified areas.
